tools/haskell-translator/caseconvs

```diff
@@ -2092,3 +2092,216 @@ case \x of Arch.Types.APIObjectType _ -> Arch.Types.SmallPageObject -> Arch.Type
+    if isPageCap a \<and> isPageCap b
+    then ->1
+    else if isPageTableCap a \<and> isPageTableCap b
+    then ->2
+    else if isPageDirectoryCap a \<and> isPageDirectoryCap b
+    then ->3
+    else if isASIDControlCap a \<and> isASIDControlCap b
+    then ->4
+    else if isASIDPoolCap a \<and> isASIDPoolCap b
+    then ->5
+    else if isSGISignalCap a \<and> isSGISignalCap b
+    then ->6
+    else ->7
+
+case \x of ((a@PageCap { capVPBasePtr = ptrA }), (b@PageCap {})) -> ((SGISignalCap {}), _) -> (a, b) ->  --->  let (a, b) = \x in
+    if isPageCap a \<and> isPageCap b
+    then let ptrA = capVPBasePtr a
+        in ->1
+    else if isSGISignalCap a
+    then ->2
+    else ->3
+
+case \x of ArchInv.InvokeSGISignal iv -> _ ->  ---> case \x of
+    InvokeSGISignal iv \<Rightarrow> ->1
+  | _ \<Rightarrow> ->2
+
+case \x of cap@(PageDirectoryCap {}) -> cap@(PageTableCap {}) -> cap@(PageCap {}) -> ASIDControlCap -> cap@(ASIDPoolCap {}) -> _ -> ---> let cap = \x in
+    if isPageDirectoryCap cap
+    then ->1
+    else if isPageTableCap cap
+    then ->2
+    else if isPageCap cap
+    then ->3
+    else if isASIDControlCap cap
+    then ->4
+    else if isASIDPoolCap cap
+    then ->5
+    else ->6
+
+case \x of (c@PageTableCap { capPTMappedAddress = Just _ }) -> (PageTableCap { capPTMappedAddress = Nothing }) -> (c@PageDirectoryCap { capPDMappedASID = Just _ }) -> (PageDirectoryCap { capPDMappedASID = Nothing }) -> (c@PageCap {}) -> c@ASIDControlCap -> (c@ASIDPoolCap {}) -> (c@VCPUCap {}) -> (c@SGISignalCap {}) ->  ---> let c = \x in
+    if isPageTableCap c \<and> capPTMappedAddress c \<noteq> None
+    then ->1
+    else if isPageTableCap c \<and> capPTMappedAddress c = None
+    then ->2
+    else if isPageDirectoryCap c \<and> capPDMappedASID c \<noteq> None
+    then ->3
+    else if isPageDirectoryCap c \<and> capPDMappedASID c = None
+    then ->4
+    else if isPageCap c
+    then ->5
+    else if isASIDControlCap c
+    then ->6
+    else if isASIDPoolCap c
+    then ->7
+    else if isVCPUCap c
+    then ->8
+    else if isSGISignalCap c
+    then ->9
+    else undefined
+
+case \x of ((a@PageCap {}), (b@PageCap {})) -> ((a@PageTableCap {}), (b@PageTableCap {})) -> ((a@PageDirectoryCap {}), (b@PageDirectoryCap {})) -> (ASIDControlCap, ASIDControlCap) -> ((a@ASIDPoolCap {}), (b@ASIDPoolCap {})) -> ((a@VCPUCap {}), (b@VCPUCap {})) -> ((a@SGISignalCap {}), (b@SGISignalCap {})) -> (_, _) ->  ---> let (a, b) = \x in
+    if isPageCap a \<and> isPageCap b
+    then ->1
+    else if isPageTableCap a \<and> isPageTableCap b
+    then ->2
+    else if isPageDirectoryCap a \<and> isPageDirectoryCap b
+    then ->3
+    else if isASIDControlCap a \<and> isASIDControlCap b
+    then ->4
+    else if isASIDPoolCap a \<and> isASIDPoolCap b
+    then ->5
+    else if isVCPUCap a \<and> isVCPUCap b
+    then ->6
+    else if isSGISignalCap a \<and> isSGISignalCap b
+    then ->7
+    else ->8
```

Comment:
That's a lot more caseconvs than I expected! Must've been painful. They look correct though.